README.md

```diff
@@ -1,30 +1,192 @@
+The inputs of the `sponge_hash` are always `Scalar` or need to be capable of being represented
+as it.
+
+The module provides two sponge hash implementations:
+- Sponge hash using `Scalar` as backend. Which hashes the inputed `Scalar`s and returns a single
+`Scalar`.
+
+- Sponge hash gadget using `dusk_plonk::Variable` as a backend. This techniqe is used/required
+when you want to proof pre-images of unconstrained data inside of Zero-Knowledge PLONK circuits.
+
+
+## Merkle Hash
+The Merkle Level Hashing is a technique that Poseidon is optimized-by-design
+to perform.
+This technique allows us to perform hashes of an entire Merkle Tree using
+`Hades252` as backend.
+
+The technique requires the computation of a `bitflags` element which is always
+positioned as the first item of the level when we hash it, and it basically generated
+in respect of the presence or absence of a leaf in the tree level.
+This allows to prevent hashing collitions.
+
+At the moment, this library is designed and optimized to work only with trees of `ARITY`
+up to 4. **That means that trees with a bigger ARITY SHOULD NEVER be used with this lib.**
+The module contains the implementation of 4 variants of the same algorithm to support the
+majority of the configurations that the user may need:
+
+- Scalar backend for hashing Merkle Tree levels outside of ZK-Circuits whith two variants:
+One of them computes the bitflags item while the other assumes that it has already been
+computed and placed in the first Level position.
+
+- `dusk_plonk::Variable` backend for hashing Merkle Tree levels inside of ZK-Circuits,
+ specifically, PLONK circuits. This implementation comes also whith two variants;
+One of them computes the bitflags item while the other assumes that it has already been
+computed and placed in the first Level position.
+
+
+
+### Zero Knowledge Merkle Opening Proof example:
+```rust
+use poseidon252::{StorageScalar, PoseidonAnnotation};
+use poseidon252::merkle_proof::merkle_opening_gadget;
+use dusk_plonk::commitment_scheme::kzg10::PublicParameters;
+use dusk_plonk::constraint_system::{Variable, StandardComposer};
+use dusk_plonk::fft::EvaluationDomain;
+use dusk_bls12_381::Scalar;
+use kelvin::{Blake2b, Compound};
+use kelvin_hamt::{HAMTSearch, NarrowHAMT};
+use merlin::Transcript;
+// Generate Composer & Public Parameters
+let pub_params = PublicParameters::setup(1 << 17, &mut rand::thread_rng()).unwrap();
+let (ck, vk) = pub_params.trim(1 << 16).unwrap();
+// Generate a tree with random scalars inside.
+let mut hamt: NarrowHAMT<_, _, PoseidonAnnotation, Blake2b> = NarrowHAMT::new();
+for i in 0..1024u64 {
+    hamt.insert(i, StorageScalar(Scalar::from(i as u64)))
+        .unwrap();
+}
+
+for i in [0u64, 567, 1023].iter() {
+    let mut composer = StandardComposer::new();
+    let mut transcript = Transcript::new(b"Test");
+    // We want to proof that we know the Scalar tight to the key Xusize
```

Comment:
tied?

Reply:
Good one!

Reply:
also either "we want to prove" or "we want proof that"